src/drivers/net/dwc_gmac/dwc_gmac.c

```diff
@@ -0,0 +1,596 @@
+
+	assert(idx < RX_DESC_QUANTITY);
+
+	desc = &rx_descs_pool[idx];
+
+	desc->basic.des2 = (uint32_t)&rx_buffers[idx][0];
+
+	if (idx < (RX_DESC_QUANTITY - 1)) {
+		next_desc = idx + 1;
+		desc->basic.des1 = ERDES1_SECOND_ADDRESS_CHAINED | ETH_FRAME_LEN;
+	} else {
+		next_desc = 0;
+		desc->basic.des1 = ERDES1_SECOND_ADDRESS_CHAINED | ETH_FRAME_LEN /*| ERDES1_END_RING */;
+	}
+
+	desc->basic.des3 = (uint32_t)&rx_descs_pool[next_desc];
+
+	dcache_flush(desc, sizeof(*desc));
+	desc->basic.des0 = RDES0_OWN;
+	dcache_flush(desc, sizeof(*desc));
+
+	return next_desc;
+}
+
+
+static inline int dwc_rxfinish_locked(struct net_device *dev_id) {
+	int cur_desc;
+	struct sk_buff *skb;
+	struct dwc_priv *priv;
+	struct dma_extended_desc *desc;
+	int len;
+
+	priv = dev_id->priv;
+
+	for(;;) {
+		cur_desc = priv->rxdesc_id;
+
+
+		desc = &((struct dma_extended_desc *)priv->rxdesc_ring_paddr)[cur_desc];
+		dcache_inval(desc, sizeof(*desc));
+
+		if (desc->basic.des0 & RDES0_OWN) {
+			return 0;
+		}
+
+		len = (desc->basic.des0 >> 16) & 0x1FFF;
+		dcache_inval((void *)desc->basic.des2, len);
+
+
+		skb = skb_alloc(len);
+		if (!skb) {
+			log_error("couldn't allocate skb");
+			return -ENOMEM;
+		}
+		memcpy(skb_data_cast_in(skb->data), (void *)desc->basic.des2, len);
+
+		skb->len = len - 4; /* without CRC */
+		skb->dev = dev_id;
+
+		show_packet(skb_data_cast_in(skb->data), skb->len, "rx");
+
+		netif_rx(skb);
+
+		dwc_setup_rxdesc(priv->rxdesc_id);
+
+		if (priv->rxdesc_id < (RX_DESC_QUANTITY - 1)) {
+			priv->rxdesc_id++;
+		} else {
+			priv->rxdesc_id = 0;
+		}
+	}
+
+	return 0;
+}
+
+static inline int dwc_txfinish_locked(struct net_device *dev_id) {
+	/* do nothing. we have been already freed skb */
+	return 0;
+}
+
+static irq_return_t dwc_irq_handler(unsigned int irq_num, void *dev_id) {
+	struct dwc_priv *priv;
+	uint32_t reg;
+	uint32_t tmp;
+	ipl_t ipl;
+
+	assert(dev_id);
+
+	priv = ((struct net_device *)dev_id)->priv;
+
+	reg = dwc_reg_read(priv, DWC_GMAC_INTERRUPT_STATUS);
+	if (reg) {
+		tmp = dwc_reg_read(priv, DWC_GMAC_SGMII_RGMII_SMII_CTRL_STATUS);
+		log_debug("DWC_GMAC_INTERRUPT_STATUS %#X", tmp);
+	}
+
+	ipl = ipl_save();
+	{
+		reg = dwc_reg_read(priv, DWC_DMA_STATUS);
+		if (reg & DWC_DMA_STATUS_NIS) {
+			if (reg & DWC_DMA_STATUS_RI) {
+				dwc_rxfinish_locked(dev_id);
+				log_debug("DWC_DMDWC_DMA_STATUS_RI");
+			}
+
+			if (reg & DWC_DMA_STATUS_TI) {
+				dwc_txfinish_locked(dev_id);
+				log_debug("DWC_DMDWC_DMA_STATUS_TI");
+			}
+		}
+
+		if (reg & DWC_DMA_STATUS_AIS) {
+			if (reg & DWC_DMA_STATUS_FBI) {
+				/* Fatal bus error */
+				log_error("Ethernet DMA error, restarting controller (%s)", ((struct net_device *)dev_id)->name);
+			}
+		}
+
+		dwc_reg_write(priv, DWC_DMA_STATUS, reg & DWC_DMA_STATUS_INTR_MASK);
+	}
+	ipl_restore(ipl);
+
+	return IRQ_HANDLED;
+}
+
+static int dwc_tx_ring_init(struct dwc_priv *priv) {
+	struct dma_extended_desc *desc;
+
+	desc = &tx_descs_pool[0];
+	memset(desc, 0, sizeof(*desc));
```

Comment:
This function sets _single_ buffer to zero. I guess that's not what you want.

It should be rather

`memset(tx_descs_pool, 0, sizeof(tx_descs_pool))`

Or set buffer to zero in `dwc_setup_txdesc`.

Reply:
>  It's enough to init first descriptor (clean OWN bit)

Let's assume that buffer is filled with garbage (i.e. `OWN` bit can be set for any buffer).

This `memset()` clears the first buffer descriptor, but second buffer may still have `OWN` bit set. When OS will try to send the first packet, following problem may occur: in `xmit() -> dwc_setup_txdesc()` second buffer is marked as next in the linked list:
```
 desc->basic.des3 = (uint32_t)&tx_descs_pool[next_desc];   
```

But `OWN` is not cleared (it still has garbage in the descriptor as we didn't clear it), so controller will try to send second buffer as well.